contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -2206,36 +2252,59 @@ static List
  * inputs are sanitized to prevent unexpected behaviour.
  */
 void
-exec_grantschema_subcmds(const char *schema, const char *rolname, bool is_grant, bool with_grant_option, AclMode privilege)
+exec_grantschema_subcmds(const char *schema, const char *rolname, bool is_grant, bool with_grant_option, AclMode privilege, bool is_create_schema)
 {
 	List		*parsetree_list;
 	ListCell	*parsetree_item;
-
-	parsetree_list = gen_grantschema_subcmds(schema, rolname, is_grant, with_grant_option, privilege);
-	/* Run all subcommands */
-	foreach(parsetree_item, parsetree_list)
+	const char *dbo_role;
+	const char *dbname = get_cur_db_name();
+	bool 			is_set_userid = false;
+	Oid 			save_userid;
+	int 			save_sec_context;
+	MigrationMode baseline_mode = is_user_database_singledb(dbname) ? SINGLE_DB : MULTI_DB;
+	dbo_role = get_dbo_role_name_by_mode(dbname, baseline_mode);
+
+	/* Need dbo user to execute the statements. */
+	PG_TRY();
+	{
+		GetUserIdAndSecContext(&save_userid, &save_sec_context);
+		SetUserIdAndSecContext(get_role_oid(dbo_role, true),
+					save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
+		is_set_userid = true;
+		parsetree_list = gen_grantschema_subcmds(schema, rolname, is_grant, with_grant_option, privilege, is_create_schema);
+		/* Run all subcommands */
+		foreach(parsetree_item, parsetree_list)
+		{
+			Node		*stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+			PlannedStmt *wrapper;
+
+			/* need to make a wrapper PlannedStmt */
+			wrapper = makeNode(PlannedStmt);
+			wrapper->commandType = CMD_UTILITY;
+			wrapper->canSetTag = false;
+			wrapper->utilityStmt = stmt;
+			wrapper->stmt_location = 0;
+			wrapper->stmt_len = 0;
+
+			/* do this step */
+			ProcessUtility(wrapper,
+						"(GRANT SCHEMA )",
+						false,
+						PROCESS_UTILITY_SUBCOMMAND,
+						NULL,
+						NULL,
+						None_Receiver,
+						NULL);
+		}
+	}
+	PG_CATCH();
```

Comment:
We should either change this to PG_FINALLY or add the missing pg_rethrow in catch.